course5/homework12/line_sorting.cc

```diff
@@ -0,0 +1,47 @@
+#include <stdexcept>
+#include <limits>
+
+#include "line_sorting.h"
+
+namespace jds {
+
+std::vector<std::string> ReadLines(std::istream & source)
+{
+	std::vector<std::string> lines;
+	std::string line;
+	while (source)
+	{
+		std::getline(source, line);
+		if (bool(source) || line.length())
+			lines.push_back(std::move(line));
```

Comment:
According to Compiler Explorer, without `std::move` here the compiler will generate a call to `std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::push_back(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)` which looks like it will trigger a copy constructor because the argument is of type `std::string const&`. With `std::move` the argument type changes to `std::string &&` to allow a move construction call.